### ISSUE TITLE
Increased the n_components to 3?

### ISSUE THREAD

**Issue by wyhhyw123** [pavlin-policar/openTSNE#121]
Hello, this is a great work and help me a lot, thanks very much. But i have some doubts that why not increased the n_components to 3? Since I participated in some competitions, I found that retaining 3 components can achieve higher accuracy using sklearn.mainfold.tsne.

**Comment by pavlin-policar** [comment]
Hi, I believe this question has already been answered in #84, but the primary reason is that it would be really slow. openTSNE implements an approximation algorithm called FIt-SNE which scales linearly in the number of points. However, it scales _exponentially_ in the number of target dimensions, meaning 3d visualizations would take exponentially longer than 2D.

In general, I don't like 3D point visualizations because there is no good way to visualize them when printing to 2D paper, or when looking at them on a 2D screen. You always need interactivity on a computer for that to make sense. Maybe it will be relevant when VR tools are developed.

> Since I participated in some competitions, I found that retaining 3 components can achieve higher accuracy using sklearn.mainfold.tsne.

I am not sure what you mean by accuracy here. The goal of t-SNE is visualization and EDA unless you're using the `transform` functionality, which should also be used for EDA. The predictive accuracy of transform will be very similar to that of a KNN model. Some packages, like scikit-learn, use the original parameters for t-SNE, which can often lead to arbitrary structure in visualizations. openTSNE and FIt-SNE both use different parameters, which drastically improve the quality of embeddings and the clusters are organized in much more meaningful ways. So if you were finding structure in 3D from scikit-learn that you aren't getting in 2D openTSNE, I would wonder if that structure is real at all.

If you really need 3D t-SNE, I believe the only implementation that allows this right now is scikit-learn. I would then suggest you use PCA initialization and use the top 3 components and increase the learning rate to N/12 (this may be too high -- this has only been tested on 2D, but it works really well there). This alone should give you much better visualizations than the scikit-learn default. scikit-learn's t-SNE is, unfortunately, one of the slowest implementations out there (even with their recent multithreading addition) so it may take a long time to run.

I have no intention of implementing this is the foreseeable future as I feel it would be a lot of effort for something that, in my opinion, is not really all that useful. I'd love to hear any use cases where 3D embeddings might be more useful and informative than 2D embeddings, but it seems to me that people like them just because they're pretty.

**Comment by dkobak** [comment]
@pavlin-policar But doesn't Barnes-Hut in openTSNE support embedding into more than 2 dimensions?

**Comment by pavlin-policar** [comment]
Oh, I suppose it does. @wyhhyw123 Use that then. It should work, and you'll get better embeddings. I don't remember ever actually testing this in 3D, I guess it was just easier writing the BH code this way. It should be faster than scikit-learn though.

**Comment by dkobak** [comment]
I don't know what happens when openTSNE is run with `n_components=3`. Maybe it should give a warning that it will switch to Barnes-Hut because FFT approximation for 3d is not implemented. Or at least suggest the user to use BH via `negative_gradient_method='bh'`.

**Comment by pavlin-policar** [comment]
Yes, adding that to the exception message would be a good idea.

**Comment by wyhhyw123** [comment]
thanks very much. I have tried using Barnes-Hut to keep 3 dimensions. It is at least 5 times faster than sklearn. Great!



---Original---

**Comment by dkobak** [comment]
> Using Barnes Hut removes the availibilty to use transform on the embeddings right?

Transform should still work.
